### PR TITLE
Move index file to flow, to allow better internal typechecking with core react-native

### PR DESCRIPTION
**Pull request by acoates-ms**
Rather than having a root Typescript file which imports the flow file and modifies the exported object, we just use the override mechanism to override the root index file, and add our specific exports.

This will allow us to move our additional components to flow and get better type checking, since many of them talk to internal RN interfaces rather than just talking to the public RN interface.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5508)

### CODE REVIEW

File: vnext/.flowconfig [hunk 1]
```diff
@@ -95,8 +93,10 @@ module.file_ext=.windows.js
 
 munge_underscores=true
 
-module.name_mapper='^react-native$' -> '<PROJECT_ROOT>/index.js'
+module.name_mapper='^react-native$' -> '<PROJECT_ROOT>/index.windows.js'
+module.name_mapper='^react-native-windows$' -> '<PROJECT_ROOT>/index.windows.js'
 module.name_mapper='^react-native/\(.*\)$' -> '<PROJECT_ROOT>/\1'
+module.name_mapper='^react-native-windows/\(.*\)$' -> '<PROJECT_ROOT>/\1'
```
Comment (aeulitz) [inline vnext/.flowconfig:99]: >.* [](start = 44, length = 2)

[nit] requires backtracking, potentially unexpected results if multiple ')'s

File: vnext/local-cli/src/runWindows/utils/msbuildtools.ts [hunk 2]
```diff
@@ -27,18 +27,18 @@ import {findLatestVsInstall} from './vsInstalls';
 export default class MSBuildTools {
   /**
    * @param version is something like 16.0 for 2019
-   * @param path  Path to MSBuild.exe (x86)
+   * @param msbuildPath  Path to MSBuild.exe (x86)
```
Comment (aeulitz) [inline vnext/local-cli/src/runWindows/utils/msbuildtools.ts:30]: >msbuildPath [](start = 12, length = 11)

OOI, what prompted the rename?

File: vnext/.flowconfig (hunk 1 above)
Comment (acoates-ms) [reply vnext/.flowconfig:99]: ?

File: vnext/local-cli/src/runWindows/utils/msbuildtools.ts (hunk 2 above)
Comment (acoates-ms) [reply vnext/local-cli/src/runWindows/utils/msbuildtools.ts:30]: nested properties called path.  Its a lint error.  Path from require(path)

File: vnext/src/index.windows.js [hunk 3]
```diff
@@ -0,0 +1,689 @@
...
+import typeof Pressable from './Libraries/Components/Pressable/Pressable';
+import typeof ProgressBarAndroid from './Libraries/Components/ProgressBarAndroid/ProgressBarAndroid';
+import typeof ProgressViewIOS from './Libraries/Components/ProgressViewIOS/ProgressViewIOS';
+import typeof SafeAreaView from './Libraries/Components/SafeAreaView/SafeAreaView';
+import typeof ScrollView from './Libraries/Components/ScrollView/ScrollView';
+import typeof SectionList from './Libraries/Lists/SectionList';
+import typeof SegmentedControlIOS from './Libraries/Components/SegmentedControlIOS/SegmentedControlIOS';
+import typeof Slider from './Libraries/Components/Slider/Slider';
+import typeof Switch from './Libraries/Components/Switch/Switch';
+import typeof RefreshControl from './Libraries/Components/RefreshControl/RefreshControl';
```
Comment (aeulitz) [inline vnext/src/index.windows.js:37]: >RefreshControl [](start = 14, length = 14)

OOI, what determines the order of these imports?

File: vnext/src/index.windows.js [hunk 4]
```diff
@@ -0,0 +1,689 @@
...
+ * @flow
+ */
+
+'use strict';
+
+import typeof AccessibilityInfo from './Libraries/Components/AccessibilityInfo/AccessibilityInfo';
+import typeof ActivityIndicator from './Libraries/Components/ActivityIndicator/ActivityIndicator';
+import typeof Button from './Libraries/Components/Button';
+import typeof CheckBox from './Libraries/Components/CheckBox/CheckBox';
+import typeof DatePickerIOS from './Libraries/Components/DatePicker/DatePickerIOS';
```
Comment (aeulitz) [inline vnext/src/index.windows.js:17]: >IOS [](start = 24, length = 3)

OOI, why are there "IOS" and "Android" imports in a .windows.js file?

File: vnext/src/index.windows.js (hunk 3 above)
Comment (acoates-ms) [reply vnext/src/index.windows.js:37]: This file is almost entirely just the base react-native one.. with a few extra imports at the end of the object.

File: packages/react-native-win32/overrides.json [hunk 5]
```diff
@@ -12,8 +12,12 @@
...
     },
     {
-      "type": "platform",
-      "file": "src/index.win32.js"
+      "type": "derived",
+      "file": "src/index.win32.js",
+      "baseFile": "index.js",
+      "baseVersion": "0.0.0-56cf99a96",
+      "baseHash": "2053998d5a0b562e08086c9b6499199ffe8ead72",
+      "issue": "LEGACY_FIXME"
```
Comment (NickGerleman) [inline packages/react-native-win32/overrides.json:20]: For derived overrides you can leave this prop off entirely if it isn't worth having an issue

File: vnext/.flowconfig (hunk 1 above)
Comment (aeulitz) [reply vnext/.flowconfig:99]: Never mind, it's nitpick irrelevant in this context. As specified, "rnw/(a))" would map to "<PRJROOT>/a)"; a  sub-pattern like "\([^\(\)]*\)" would map to "<PRJROOT>/a". 

---
In reply to: [453108200](https://github.com/microsoft/react-native-windows/pull/5508#discussion_r453108200) [](ancestors = 453108200)

File: vnext/src/index.windows.js (hunk 3 above)
Comment (TheSavior) [reply vnext/src/index.windows.js:37]: I think the theory was Components alphabetical, then APIs alphabetical. *shrug*. I believe it is the same order as the exports

File: packages/react-native-win32/overrides.json (hunk 5 above)
Comment (acoates-ms) [reply packages/react-native-win32/overrides.json:20]: Ah.  I agree with it not being worth having an issue -- hence why I didn't create one.

**Pull request merged**